packages/compiler/auto.ts

```diff
@@ -0,0 +1,306 @@
+
+  const rawComponent = path.node;
+
+  const isInProgramScope =
+    t.isIdentifier(rawComponent.id) &&
+    info.programPath.scope.hasOwnBinding(rawComponent.id.name);
+
+  if (!isInProgramScope) return;
+
+  if (
+    t.isCallExpression(path.parent) &&
+    t.isIdentifier(path.parent.callee, { name: 'block' }) // probably need a import level check later
+  )
+    return;
+
+  let componentPath: NodePath<
+    t.FunctionDeclaration | t.FunctionExpression | t.ArrowFunctionExpression
+  >;
+
+  // Next.js React Server Components support
+  if (typeof info.options.auto === 'object' && info.options.auto.rsc) {
+    const directives = info.programPath.node.directives;
+    if (!directives.length) return; // we assume it's server component only
+    const hasUseClientDirective = directives.some((directive) => {
+      return directive.value.value === 'use client';
+    });
+    if (!hasUseClientDirective) return;
+  }
+
+  if (t.isVariableDeclarator(rawComponent)) {
+    if (
+      !t.isArrowFunctionExpression(rawComponent.init) &&
+      !t.isFunctionExpression(rawComponent.init)
+    ) {
+      return;
+    }
+    // it's possible that the component is a block
+    // probably need a import level check later)
+    if (
+      t.isCallExpression(rawComponent.init) &&
+      t.isIdentifier(rawComponent.init, { name: 'block' })
+    )
+      return;
+    if (rawComponent.init.async) return; // RSC / Loader
+
+    componentPath = path.get('init') as NodePath<
+      t.FunctionExpression | t.ArrowFunctionExpression
+    >;
+  } else {
+    componentPath = path as NodePath<t.FunctionDeclaration>;
+  }
+
+  if (!t.isIdentifier(rawComponent.id)) return;
+  if (!isComponent(rawComponent.id.name)) return;
+
+  const globalPath = path.findParent(
+    (path) => path.parentPath?.isProgram() || path.isProgram()
+  )!;
+  const comment =
+    rawComponent.leadingComments?.[0] ||
+    path.parent.leadingComments?.[0] ||
+    globalPath.node.leadingComments?.[0];
+
+  if (comment?.value.includes('million-ignore')) {
+    logIgnore(rawComponent.id.name);
+    return;
+  }
+
+  const blockStatementPath = componentPath.get(
+    'body'
+  ) as NodePath<t.BlockStatement>;
+  const returnStatementPath = blockStatementPath
+    .get('body')
+    .find((path) => path.isReturnStatement());
+
+  if (!returnStatementPath?.has('argument')) return;
+  const argumentPath = returnStatementPath.get('argument') as NodePath<
+    t.JSXElement | t.JSXFragment
+  >;
+
+  if (!argumentPath.isJSXElement() && !argumentPath.isJSXFragment()) {
+    return;
+  }
+
+  const componentInfo = {
+    bailout: false,
+    elements: 0,
+    attributes: 0,
+    components: 0,
+    text: 0,
+    expressions: 0,
+    depth: 0,
+    returns: 0,
+  };
+
+  const skipIds: (string | RegExp)[] = [];
+
+  if (typeof info.options.auto === 'object' && info.options.auto.skip) {
+    skipIds.push(...info.options.auto.skip);
+  }
+
+  blockStatementPath.traverse({
+    Identifier(path) {
```

Comment:
Be careful on this one, as this covers non-expression identifiers like object keys. I would recommend traversing `Expression` instead and then checking if the node is an `Identifier`